### DISCUSSION TITLE
Weird force curve shapes

### DISCUSSION THREAD

**Discussion by Gordon-Shumway2** [JaapvanEkris/openrowingmonitor#38] [General]
Sanity check, please.

I am getting these strange force curve shapes, full of bumps and troughs on my water-based rower: 

![vlcsnap-2024-02-08-19h01m05s333](https://github.com/JaapvanEkris/openrowingmonitor/assets/78542107/b7cfd61c-728a-46c2-bf57-34f69d2a4625)
![vlcsnap-2024-02-08-19h01m34s902](https://github.com/JaapvanEkris/openrowingmonitor/assets/78542107/49e4f916-d56f-4e08-899d-2f607c8f8519)
![vlcsnap-2024-02-08-19h02m01s160](https://github.com/JaapvanEkris/openrowingmonitor/assets/78542107/9e5a13f9-0dc5-4178-b857-3677f9e6d10b)

Could someone confirn that they should not look like that? I know that I do not have the best technique but it is not that bad....
I suspect that there is a mechanical issue with a ball bearing that is in the power transmission path between handle and water tank.

Cheers,
Michael aka Gordon

**Comment by JaapvanEkris** [comment]
Could this be a software thing? I changed some of the algorithms, so some data may behave differently. Or do you suspect a hardware error (noises, weird resistance in the machine)?

**Reply by Gordon-Shumway2** [reply]
I don't think it is software - I get similar shapes when I run the same data through rowsandall's force curve plot. So the weirdness seems to be in the data (assuming that your algorithm changes deal with plotting/presenting the data, not with gathering it),

There seem to be 3-4 dips in the curve which happens to coincide with the number of rotations during a drive for both the water wheel and a deflection roller. That deflection roller has two ball bearings. I just took the roller out of the rower's metal frame and indeed one of the two ball bearings has quite a bit more friction than the other. Will open a case with the manufacturer to get a replacement.

**Reply by JaapvanEkris** [reply]
Well, I improved the algorithm to calculate flywheel velocity and acceleration, as I had a small breakthrough in reducing the CPU load caused by it (sometimes over 50% reduction). What I have seen so far is that force curves should become smoother.

**Reply by Abasz** [reply]
@Gordon-Shumway2 can you send the delta times csv for this row? I will run it through the version prior to Jaaps improvement to see  what kind of force curves that produces.

Also kindly send me your config.js (or other files if you changed them) so I can run the simulation identical to your setup.

**Reply by Gordon-Shumway2** [reply]
Hello Abasz, 

thank you for looking into this. Please do not spend too much time on it, I am increasingly convinced this is a hardware issue. 

Please find the .csv  and `config.js` and `rowerprofiles.js` in this response - file extensions were modified from ".js" to ".json" to allow upload to this thread.

[config.json](https://github.com/JaapvanEkris/openrowingmonitor/files/14232417/config.json)
[rowerProfiles.json](https://github.com/JaapvanEkris/openrowingmonitor/files/14232418/rowerProfiles.json)
[2024-02-07_15-32-41_rowingData.csv](https://github.com/JaapvanEkris/openrowingmonitor/files/14232422/2024-02-07_15-32-41_rowingData.csv)


Kind regards, 
Michael

**Reply by Abasz** [reply]
Sorry I meant the raw delta times so I can simulate your session with older version (i.e. old detection) :)

THanks

**Reply by Gordon-Shumway2** [reply]
OK, I did a new short rowing session which shows similarly weird force curve shapes and recorded the raw csv data.
[2024-02-12_11-30-39_raw.csv.gz](https://github.com/JaapvanEkris/openrowingmonitor/files/14237841/2024-02-12_11-30-39_raw.csv.gz)


Thanks again for looking into this, 
Michael

**Reply by Abasz** [reply]
I have re run your session with both old and new ORM The results are quite obvious. The reason I took more interest in this as I have observed something similar but less severe. I theorise that the spikes as you have less data point on your water rower than on my air rower is the cause of the data points not visisble. Below is a compare of the before after of a stroke where I observed the same behaviour:
weightedAVG 
![image](https://github.com/JaapvanEkris/openrowingmonitor/assets/32517724/0e96f156-c5ce-416a-91f9-08901ace3277)

Median: 
![image](https://github.com/JaapvanEkris/openrowingmonitor/assets/32517724/8f677565-f90d-4e9f-8584-20db468c6904)

combined
![image](https://github.com/JaapvanEkris/openrowingmonitor/assets/32517724/89613326-9940-499f-a3a3-945461fc3dae)

 
It is clear that the impulses are not consistent and there is noise during the recovery (hence the tail of the curve is choppy in the median version). With the weighted average these spikes are smoothed out but some spikes (though smoothed out) still remain. I dont think this can be effectively filtered as it comes from the machine it self.

I have run your provided data and here is the before after (blue is the new and orange is the old algorithm):
![image](https://github.com/JaapvanEkris/openrowingmonitor/assets/32517724/be813b3f-d137-4539-b486-370b24aa7869)

So this is not an OrM issue. Water rowers do not have sufficient data points (even if you have more mangets) for the drive phase really not to mention that due to the variable resistance from the water it results in not so clean delta time impulses. Also I can see that you have 6 magnets per rotation, which means that there is additional noise from the magnet placements. This cumulative effect of noise results in poor force curves. 

Now it was slightly interesting that looking at your full data set the signals looked rather clean:

![image](https://github.com/JaapvanEkris/openrowingmonitor/assets/32517724/984709e1-0bf6-4960-b64e-080ca3e6bc50)

But I suspect that the end of your drive phase (where the resistance of the water and your pull force is almost equalise) becomes choppy due to the variable resistance of the water as it moves in the tank. I mean if you look at it the beginning of your drive (where you pull the most) consist of basically 2 datapoints.


EDIT:

I have been looking at your settings and it seems that you have 6 datapoints per rotation, yet you set the flanklngth to 3. Are you sure you have 6 impulses per one rotation and not one?

Or if you really have 6, why have you settled with flankLength of 3? I know that the 1.5-2 rotations for the flanklenght (so 9-12) is only for air rower but I think 3 for 6 data points would still be quite low.

**Reply by JaapvanEkris** [reply]
An additional note: water rowers typically have low rotational speeds than air rowers. The waterbased "flywheel" by nature is heavier, as is its inertia, but it can't rotate as fast as a air based flywheel. A typical air based flywheel will hit 1000+ RPM....

**Reply by Abasz** [reply]
> An additional note: water rowers typically have low rotational speeds than air rowers. The waterbased "flywheel" by nature is heavier, as is its inertia, but it can't rotate as fast as a air based flywheel. A typical air based flywheel will hit 1000+ RPM....

Thanks, you are right the 1.5-2 rotation recommendation is for air rower. Edited my post to avoid confusion

**Comment by Abasz** [comment]
On an other note, can you send me the resolution (size and pixel ratio, or the type of the screen it selfe) of the screen you are running ORM? I dont like that the function icons dont line up properly... I have tested some resolutions but these PI screens have random sizes as pixel ratio.... :)

**Reply by Gordon-Shumway2** [reply]
These screen shots were taken from Firefox on Win10 - I cropped the images to remove some personal info from browser tabs. The resolution was set to 1920 x 1080.

**Reply by Abasz** [reply]
Aaa ok, I never tested the layout on so high Res:) would never thought (probably naively) that these will be used with small screens ')

**Comment by Gordon-Shumway2** [comment]
Update: found a solution (or rather @Abasz did)

After reading the thread at https://github.com/JaapvanEkris/openrowingmonitor/discussions/42 which mentions modifying _flankLength_ I increased _flankLength_ from 3 to 5 and got nice force curves without major bumps and troughs!

Thank you!

**Reply by Abasz** [reply]
Good to hear:) though we've already suggested this just not explicitly that is true:)

> Or if you really have 6, why have you settled with flankLength of 3? I know that the 1.5-2 rotations for the flanklenght (so 9-12) is only for air rower but I think 3 for 6 data points would still be quite low.
